crates/top/re_sdk/src/lenses/ast.rs

```diff
@@ -4,49 +4,92 @@
 
-use std::collections::HashMap;
+use std::sync::Arc;
 
 use arrow::{
-    array::{Int64Array, ListArray},
-    datatypes::DataType,
+    array::{AsArray as _, Int64Array, ListArray},
+    compute::take,
+    datatypes::{DataType, Field},
 };
+use nohash_hasher::IntMap;
 
-use re_arrow_combinators::{Transform as _, reshape::Flatten};
+use re_arrow_combinators::{
+    Transform as _,
+    reshape::{Explode, Flatten},
+};
 use re_chunk::{
-    ArrowArray as _, Chunk, ChunkComponents, ChunkId, ComponentIdentifier, EntityPath, Timeline,
+    ArrowArray as _, Chunk, ChunkId, ComponentIdentifier, EntityPath, TimeColumn, Timeline,
     TimelineName,
 };
 use re_log_types::{EntityPathFilter, TimeType};
 use re_types::{ComponentDescriptor, SerializedComponentColumn};
 
-use super::{Error, op};
+use super::{Error, builder::LensBuilder, op};
 
 pub struct InputColumn {
     pub entity_path_filter: EntityPathFilter,
     pub component: ComponentIdentifier,
 }
 
-pub struct ComponentColumn {
-    /// The target entity path for this column.
-    ///
-    /// If `None`, the entity path of the output column will be set to the matched entity path of the input column.
-    pub new_entity_path: Option<EntityPath>,
+/// Target entity path for lens outputs.
+#[derive(Debug, Clone, Default)]
+pub enum TargetEntity {
+    /// Use the matched input entity path.
+    #[default]
+    SameAsInput,
+
+    /// Use a specific entity path.
+    Explicit(EntityPath),
+}
+
+/// A component output.
+///
+/// Depending on the context in which this output is used, the result from
+/// applying the `ops` should be a list array (1:1) or a list array of list arrays (1:N).
+#[derive(Debug)]
+pub struct ComponentOutput {
     pub component_descr: ComponentDescriptor,
     pub ops: Vec<Op>,
-    // TODO(grtlr): It would be much nicer if static could be inferred from the output of the operations?
-    pub is_static: bool,
 }
 
-pub struct TimeColumn {
-    /// The target entity path for this column.
-    ///
-    /// If `None`, the entity path of the output column will be set to the matched entity path of the input column.
-    pub new_entity_path: Option<EntityPath>,
+/// A time extraction output.
+#[derive(Debug)]
+pub struct TimeOutput {
     pub timeline_name: TimelineName,
     pub timeline_type: TimeType,
     pub ops: Vec<Op>,
 }
 
+/// Determines how a lens transforms input rows to output rows.
+#[derive(Debug)]
+pub enum LensKind {
+    /// Each input row produces exactly one output row (1:1 mapping).
+    ///
+    /// Outputs inherit timelines from the input chunk.
+    OneToOne {
+        target_entity: TargetEntity,
+        components: Vec<ComponentOutput>,
+        timelines: Vec<TimeOutput>,
+    },
+
+    /// Each input row produces multiple output rows (1:N flat-map).
+    ///
+    /// Outputs are always temporal.
+    ToMany {
+        target_entity: TargetEntity,
+        components: Vec<ComponentOutput>,
+        timelines: Vec<TimeOutput>,
```

Comment:
Does this mean `timelines` is never empty? Should we maybe use [`vec1`](https://crates.io/crates/vec1) for it?